sklearn/linear_model/tests/test_logistic.py

```diff
@@ -1368,32 +1368,39 @@ def test_dtype_match(solver, multi_class):
     X_64 = np.array(X).astype(np.float64)
     y_64 = np.array(Y1).astype(np.float64)
     X_sparse_32 = sp.csr_matrix(X, dtype=np.float32)
+    solver_tol = 1e-5
 
+    lr_templ = LogisticRegression(
+        solver=solver, multi_class=multi_class,
+        random_state=42, tol=solver_tol, fit_intercept=True)
     # Check type consistency
-    lr_32 = LogisticRegression(solver=solver, multi_class=multi_class,
-                               random_state=42)
+    lr_32 = clone(lr_templ)
     lr_32.fit(X_32, y_32)
     assert_equal(lr_32.coef_.dtype, X_32.dtype)
 
     # check consistency with sparsity
-    lr_32_sparse = LogisticRegression(solver=solver,
-                                      multi_class=multi_class,
-                                      random_state=42)
+    lr_32_sparse = clone(lr_templ)
     lr_32_sparse.fit(X_sparse_32, y_32)
     assert_equal(lr_32_sparse.coef_.dtype, X_sparse_32.dtype)
 
     # Check accuracy consistency
-    lr_64 = LogisticRegression(solver=solver, multi_class=multi_class,
-                               random_state=42)
+    lr_64 = clone(lr_templ)
     lr_64.fit(X_64, y_64)
     assert_equal(lr_64.coef_.dtype, X_64.dtype)
 
-    rtol = 1e-6
+    # solver_tol bounds the norm of the loss gradient
+    # dw ~= inv(H)*grad ==>  |dw| ~= |inv(H)| * solver_tol, where H - hessian
+    #
+    # with  Z = np.hstack((np.ones((3,1)), np.array(X)))
+    # In [18]: np.linalg.norm(np.linalg.inv(0.25 * Z.T @ Z))
+    # Out[18]: 41.32740565066648
```

Comment:
What is this supposed to explain?

Reply:
Ok so `0.25 * Z.T @ Z` is the Hessian for a model that outputs p(y_k|x) = 0.5 for each class for a binary classification problem. But at optimum, it's unlikely that p(y_k|x) is close to 1.

Reply:
I am afraid my mistake here is to not take into account the regularization term.

```
In [7]: np.linalg.norm(np.linalg.inv( np.diag([0, 2, 2]) + 0.25 * Z.T @ Z))
Out[7]: 1.7193336918135917
```

this is consistent with output of the following simulation:

```python
import sklearn, numpy as np, sklearn.linear_model
import sklearn.linear_model.logistic as logistic
from sklearn.preprocessing import LabelBinarizer

X = np.array([[-1, 0], [0, 1], [1, 1]], dtype=np.double)
Y1 = np.array([0, 1, 1], dtype='i')

Y_multi = LabelBinarizer().fit_transform(Y1)
Y_multi = np.hstack([1 - Y_multi, Y_multi])

lr = sklearn.linear_model.LogisticRegression(
        solver='lbfgs', multi_class='multinomial',
        random_state=42, tol=0., max_iter=1000, 
        fit_intercept=True)
lr.fit(X, Y1)

# Optimal value obtained by minimizing in extended precision and coercing
w_exact = np.array([0.47518386020515844474, 0.36183930896547519212, 0.19148890867094416443])

def grad(w_pos):
    w = np.hstack((-w_pos, w_pos))
    C = 1.0
    alpha = 1/C
    val, grad, p = logistic._multinomial_loss_grad(w, X, Y_multi, alpha, np.ones_like(Y1))
    return grad


w_pos_fitted = np.hstack((lr.coef_, lr.intercept_.reshape((-1,1)))).ravel()

def random_unit_vectors(dim, size=1):
    x = np.random.randn(size, dim)
    x /= np.linalg.norm(x, ord=2, axis=1, keepdims=True)
    return x

def norm_inf(u):
    return np.linalg.norm(u, ord=np.inf)

print(norm_inf(grad(w_exact)))    # 1.6653345369377348e-16 

solver_tol = 1e-5
print(np.max([
   norm_inf(grad(w_exact + solver_tol*ui)) for ui in random_unit_vectors(3, 10**4)
]))

# 1.695041656418761e-05 ~ 1.695 * solver_tol
```

which gives a close constant of `1.695 * solver_tol`.

 It is odd that norm of gradient on the solution is actually much higher than the value of `tol` provided to `lbfgs` solver:

```python
print(norm_inf(grad(w_pos_fitted))) # 2.88807176396233e-07
```